[dataflow] Parse formulas from text

sam-mccall opened the pull request:
My immediate use for this is not in checked-in code, but rather the
ability to plug printed flow conditions (from analysis logs) back into
sat solver unittests to reproduce slowness.

It does allow simplifying some of the existing solver tests, though.

sam-mccall committed:
[dataflow] Parse formulas from text

My immediate use for this is not in checked-in code, but rather the
ability to plug printed flow conditions (from analysis logs) back into
sat solver unittests to reproduce slowness.

It does allow simplifying some of the existing solver tests, though.

sam-mccall commented:
Not sure if this is really important to have checked in, but I've wanted to investigate/reproduce flow condition slowness a few times now, so thought I should send the patch in case others think it's useful.

ymand commented on clang/include/clang/Analysis/FlowSensitive/Formula.h:
```diff
@@ -87,6 +87,9 @@ class alignas(const Formula *) Formula {
                          ArrayRef<const Formula *> Operands,
                          unsigned Value = 0);
 
+  // Parse Formulas using Arena rather than caling this function directly.
```
nit: Maybe prefix with "note:" or somesuch? I started reading this as a description of the function.

Xazax-hun commented on clang/lib/Analysis/FlowSensitive/Arena.cpp:
```diff
@@ -95,4 +98,94 @@ BoolValue &Arena::makeBoolValue(const Formula &F) {
+      return nullptr;
+
+    EatWhitespace();
+    if (!In.consume_front(")"))
```
Not required for this PR, but it would be great if it could do basic diagnostics like `')' expected at post Y`.

sam-mccall replied:
This makes sense, but usefully characterizing errors adds complexity - I'd like to see if this is really needed first.
(The error handling here was sufficient for the dumb mistakes I made when writing testcases...)

Xazax-hun commented on clang/lib/Analysis/FlowSensitive/Arena.cpp:
```diff
@@ -95,4 +98,94 @@ BoolValue &Arena::makeBoolValue(const Formula &F) {
+    return &(A.*Op)(*Arg1, *Arg2);
+  }
+
+  if (In.consume_front("V")) {
```
Isn't this naming convention a bit too restrictive? E.g., is it possible someone would want to more descriptive names for a test that is potentially checked in? Or maybe we never expect people to use this for regression tests?

sam-mccall replied:
Agree it's too restrictive, but solving it is a slightly bigger scope and I'm not sure of the details.

I do think it's important that if you parse a formula that uses certain names for the variables, then you can print it out again with the same names. (Parsing is mostly useful when formulas are large and have many variables...). `V0` is the default name of `Atom{0}` etc, so these variables "round-trip" OK.

To round-trip other variables we need to store their names somewhere. A map output-param works (and we have support for printing this way) but it's not very ergonomic, and is easy to forget to plumb around.
I've been thinking variable names should live on the arena (`Arena::makeAtom() -> Atom`, `Arena::nameAtom(Atom, Twine)` or so). This would provide a fairly centralized and natural way to e.g. name flow condition vars. And `parse()` could create an atom and name it when it encounters an unknown variable name.

However there's some stuff to work out here: is name tracking always on? do we also want to name values/locations, which might help provide atom names? what about Formula's `operator<<` which won't find the names?

Added a `FIXME` to handle other names.

Xazax-hun reviewed (commented):
I think it is useful to have something like this checked in, I support this PR.

sam-mccall commented on clang/include/clang/Analysis/FlowSensitive/Formula.h:
```diff
@@ -87,6 +87,9 @@ class alignas(const Formula *) Formula {
                          unsigned Value = 0);
 
+  // Parse Formulas using Arena rather than caling this function directly.
+  static Formula *parse(llvm::BumpPtrAllocator &Alloc, llvm::StringRef &);
```
Right, oops!

sam-mccall commented on clang/lib/Analysis/FlowSensitive/Arena.cpp:
```diff
@@ -95,4 +98,94 @@ BoolValue &Arena::makeBoolValue(const Formula &F) {
 
+namespace {
+const Formula *parse(Arena &A, llvm::StringRef &In) {
+  auto EatWhitespace = [&] { In = In.ltrim(' '); };
```
Renamed, hard to know if linebreaks etc are more likely to be formatting or errors at this point.

sam-mccall commented on clang/unittests/Analysis/FlowSensitive/SolverTest.cpp:
```diff
@@ -258,114 +276,71 @@ TEST(SolverTest, IffWithUnits) {
-  EXPECT_EQ(solver.solve({A, B}).getStatus(), Solver::Result::Status::TimedOut);
+  Arena A;
+  auto Constraints = parseAll(A, R"(
+     (V0 => V1)
+     (V0 & !V1)
```
I'm not changing the formulas here - the test above uses three clauses (both before & after), and this one used two.

sam-mccall committed:
Address review comments

Drop leftover parse() decl from header.
EatWhitespace -> EatSpaces
FIXME about general var names
Style nits

sam-mccall committed:
Revert accidental header inclusion
